### PR TITLE
Persist DartCallbackCache contents across launches

### PR DESCRIPTION
**Pull request by bkonyi**
Updated DartCallbackCache to write callback cache to disk which is restored on engine startup

### CODE REVIEW

File: shell/platform/android/flutter_main.cc [hunk 1]
```diff
@@ -44,7 +45,8 @@ void FlutterMain::Init(JNIEnv* env,
                        jclass clazz,
                        jobject context,
                        jobjectArray jargs,
-                       jstring bundlePath) {
+                       jstring bundlePath,
+                       jstring appRootPath) {
```
Comment (zanderso) [inline shell/platform/android/flutter_main.cc]: Who has write access to `appRootPath`? Could application code accidentally clobber the callback cache?

Comment (bkonyi) [reply shell/platform/android/flutter_main.cc]: This parameter is only visible to the internals of `FlutterMain.java` which only sets it to the result of `PathUtils.getFilesDir(applicationContext);` so it shouldn't be possible for the application to clobber the cache unless it's really trying to (e.g., by deleting the cache file directly).

File: shell/platform/android/io/flutter/view/FlutterMain.java [hunk 2]
```diff
@@ -229,7 +230,7 @@ public static void ensureInitializationComplete(Context applicationContext, Stri
         }
     }
 
-    private static native void nativeInit(Context context, String[] args, String bundlePath);
+    private static native void nativeInit(Context context, String[] args, String bundlePath, String appRootPath);
```
[the remark this answers is not in the document]
Comment (bkonyi) [reply shell/platform/android/io/flutter/view/FlutterMain.java]: SGTM, will do.

File: lib/ui/plugins/callback_cache.cc [hunk 3]
```diff
@@ -34,6 +56,7 @@ int64_t DartCallbackCache::GetCallbackHandle(const std::string& name,
 
   if (cache_.find(hash) == cache_.end()) {
     cache_[hash] = {name, class_name, library_path};
+    SaveCacheToDisk();
```
[the remark this answers is not in the document]
Comment (bkonyi) [reply lib/ui/plugins/callback_cache.cc]: Currently the cache is persistent and entries aren't deleted, although we could verify all the callbacks in the cache actually exist when we restore and remove the stale entries.

**Pull request merged**